cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,409 @@
+
+However, this is still a security-sensitive area, since we are allocating resources based on external, untrusted packets. We will need a sane circuit-breaker to ensure we do not allocate a large number of identities based on malicious input.
+
+The maximum number of identities allocated by ToFQDN selectors is the number of unique selector combinations seen for given IPs. The theoretical maximum is, therefore, $2^N-1$[^1]. 10 distinct selectors would be OK (1024). More than 16 selectors could *theoretically* exceed identity space.
+
+We need a good circuit breaker for this that does not break the S3 case. 
+
+### Key question: upgrades & downgrades?
+
+#### Background:
+It is desired that the (ip -> nid) and (selector -> nid) mappings remain stable across restarts. This is because policy updates are not completely atomic, so identity stablity is required to prevent spurious drops during the regeneration process. The window for drops widens when Envoy is also in use. This mechanism is best-effort but generally quite reliable.
+
+#### Upgrade:
+
+On restart, there will be no local allocator checkpoint, so no local identities will exist. The restoration will dump the list of (prefix -> nid) pairs, then insert a `(cidr:xxx/32)` label in to the ipcache and reserve the nid for that prefix as a "fallback"
+
+The FQDN state will be restored, loading (domain -> ip) pairs. The the (selector -> ip) mapping will be recreated, and `(fqdn:selectorXXX)` labels will be inserted in to the ipcache for relevant prefixes.
+
+After reading all k8s objects, the ipcache will proceed with injection. Every restored prefix will have the labels `(cidr:xxx/32, fqdn:selectorXXX, fqdn:selectorYYY)`. An identity will be allocated for every IP, and thanks to the reservation mechanism, the numeric ID will be stable. Thus, the selector->nids and ip->nid mappings will remain stable on upgrade, and no traffic will be dropped.
+
+#### Upgrades from v1.16 to a future version:
+
+This relies on IPs having the same set of labels before and after restart, which is not guaranteed. Logic changes may cause label sets to shift. We may need to have an "upgrade" mechanism that modifies
+existing labels based on certain upgrade heuristics.
+
+
+#### Downgrade:
+
+The constraints for downgrades are somewhat reduced; it is acceptable for brief policy drops while the agent finishes converging.
+
+On downgrade to v1.15, the restoration logic will, again, upsert (cidr:xxx/32) labels and reserved numeric identities in to the ipcache. However, for v1.16 more than one IP may share the same local identity. Thus, one IP will "capture" the numeric ID, while the others will change. Connections to those IPs may be briefly interrupted only during endpoint regeneration. Once regeneration is complete, there should be no further disruption.
+
+As a precautionary measure, v1.15.tip should be updated 
```

Comment:
Updated how?

Reply:
I've removed this, since I don't think this is necessary.

However, I've added a note to in the upgrade section that users need to upgrade to v1.15.tip for dropless upgrades (c.f. https://github.com/cilium/cilium/pull/32872)